src/cockpit/superuser.py

```diff
@@ -0,0 +1,215 @@
+# This file is part of Cockpit.
+#
+# Copyright (C) 2022 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+import asyncio
+import logging
+import os
+import pwd
+
+from systemd_ctypes import bus
+
+from typing import Dict, Optional
+
+from .router import Router, RoutingError, RoutingRule
+from .peer import Peer, PeerStateListener
+
+logger = logging.getLogger(__name__)
+
+SUPERUSER_AUTH_COOKIE = 'supermarius'
+
+# Keep a static list, for the time being.
+SUPERUSER_BRIDGES = {
+    'sudo': (
+        ['sudo', '--askpass', 'cockpit-bridge', '--privileged'], {
+            'SUDO_ASKPASS': '/usr/libexec/cockpit-askpass'
+        }
+    ),
+}
+
+
+class SuperuserStartup:
+    peer: Peer  # the peer that's being started
+
+    def success(self, rule: SuperuserRoutingRule) -> None:
+        raise NotImplementedError
+
+    def failed(self, rule: SuperuserRoutingRule, exc: Exception) -> None:
+        raise NotImplementedError
+
+    def auth(self, rule: SuperuserRoutingRule, prompt: str, echo: bool) -> None:
+        raise NotImplementedError
+
+
+class ControlMessageStartup(SuperuserStartup):
+    def success(self, rule: SuperuserRoutingRule) -> None:
+        rule.router.write_control(command='superuser-init-done')
+
+    def failed(self, rule: SuperuserRoutingRule, exc: Exception) -> None:
+        pass  # what do we do for this?
```

Comment:
I think we also send `super-user-init-done`.  Cockpit-ws doesn't care whether superuser startup has worked or not.

Reply:
The main point of super-user-init-done is to know that no more auth messages will come and the password can be removed from memory.

Reply:
Ah.  Thanks for flagging this.  I'll add it.

Reply:
Done.